controller_manager: # merged from: https://github.com/UniversalRobots/Universal_Robots_ROS2_Gazebo_Simulation/tree/ros2
  ros__parameters:
    use_sim_time: true
    update_rate: 200

    joint_state_broadcaster:
      type: joint_state_broadcaster/JointStateBroadcaster

    robotnik_base_control:
      type: diff_drive_controller/DiffDriveController
    
    # io_and_status_controller:
    #   type: ur_controllers/GPIOController

    # speed_scaling_state_broadcaster:
    #   type: ur_controllers/SpeedScalingStateBroadcaster

    # force_torque_sensor_broadcaster:
    #   type: ur_controllers/ForceTorqueStateBroadcaster

    joint_trajectory_controller:
      type: joint_trajectory_controller/JointTrajectoryController

    # scaled_joint_trajectory_controller:
    #   type: ur_controllers/ScaledJointTrajectoryController

    # forward_velocity_controller:
    #   type: velocity_controllers/JointGroupVelocityController

    # forward_position_controller:
    #   type: position_controllers/JointGroupPositionController


joint_state_broadcaster:
  ros__parameters:
    use_sim_time: true
    publish_rate: 100.0

robotnik_base_control:
  ros__parameters:
    left_wheel_names:
      - front_left_wheel_joint
      - rear_left_wheel_joint
    right_wheel_names:
      - front_right_wheel_joint
      - rear_right_wheel_joint
    wheel_separation: 0.419
    wheel_radius: 0.11
    wheel_separation_multiplier: 1.0
    left_wheel_radius_multiplier: 1.0
    right_wheel_radius_multiplier: 1.0
    odom_frame_id: odom
    base_frame_id: base_footprint
    enable_odom_tf: false
    cmd_vel_timeout: 0.1
    publish_rate: 50.0
    linear:
      x:
        has_velocity_limits: true
        max_velocity: 1.5
        has_acceleration_limits: true
        max_acceleration: 2.5
        has_jerk_limits: false
        max_jerk: 0.0
    angular:
      z:
        has_velocity_limits: true
        max_velocity: 3.0
        has_acceleration_limits: true
        max_acceleration: 2.5
        has_jerk_limits: false
        max_jerk: 0.0

# speed_scaling_state_broadcaster:
#   ros__parameters:
#     state_publish_rate: 100.0


# force_torque_sensor_broadcaster:
#   ros__parameters:
#     sensor_name: tcp_fts_sensor
#     state_interface_names:
#       - force.x
#       - force.y
#       - force.z
#       - torque.x
#       - torque.y
#       - torque.z
#     frame_id: tool0
#     topic_name: ft_data

joint_trajectory_controller:
  ros__parameters:
    joints:
      - arm_shoulder_pan_joint
      - arm_shoulder_lift_joint
      - arm_elbow_joint
      - arm_wrist_1_joint
      - arm_wrist_2_joint
      - arm_wrist_3_joint
    command_interfaces:
      - position
    state_interfaces:
      - position
      - velocity
    state_publish_rate: 100.0
    action_monitor_rate: 20.0
    allow_partial_joints_goal: false
    constraints:
      stopped_velocity_tolerance: 0.2
      goal_time: 0.0
      arm_shoulder_pan_joint: { trajectory: 0.2, goal: 0.1 }
      arm_shoulder_lift_joint: { trajectory: 0.2, goal: 0.1 }
      arm_elbow_joint: { trajectory: 0.2, goal: 0.1 }
      arm_wrist_1_joint: { trajectory: 0.2, goal: 0.1 }
      arm_wrist_2_joint: { trajectory: 0.2, goal: 0.1 }
      arm_wrist_3_joint: { trajectory: 0.2, goal: 0.1 }


# scaled_joint_trajectory_controller:
#   ros__parameters:
#     joints:
#       - arm_shoulder_pan_joint
#       - arm_shoulder_lift_joint
#       - arm_elbow_joint
#       - arm_wrist_1_joint
#       - arm_wrist_2_joint
#       - arm_wrist_3_joint
#     command_interfaces:
#       - position
#     state_interfaces:
#       - position
#       - velocity
#     state_publish_rate: 100.0
#     action_monitor_rate: 20.0
#     allow_partial_joints_goal: false
#     constraints:
#       stopped_velocity_tolerance: 0.2
#       goal_time: 0.0
#       shoulder_pan_joint: { trajectory: 0.2, goal: 0.1 }
#       shoulder_lift_joint: { trajectory: 0.2, goal: 0.1 }
#       elbow_joint: { trajectory: 0.2, goal: 0.1 }
#       wrist_1_joint: { trajectory: 0.2, goal: 0.1 }
#       wrist_2_joint: { trajectory: 0.2, goal: 0.1 }
#       wrist_3_joint: { trajectory: 0.2, goal: 0.1 }

# forward_velocity_controller:
#   ros__parameters:
#     joints:
#       - shoulder_pan_joint
#       - shoulder_lift_joint
#       - elbow_joint
#       - wrist_1_joint
#       - wrist_2_joint
#       - wrist_3_joint
#     interface_name: velocity

# forward_position_controller:
#   ros__parameters:
#     joints:
#       - shoulder_pan_joint
#       - shoulder_lift_joint
#       - elbow_joint
#       - wrist_1_joint
#       - wrist_2_joint
#       - wrist_3_joint
